File: frontend/features/submission/services/submissionService.ts
import api from '@/lib/services/axios-instance';
import { ApiResponse } from '@/lib/types';
import {
    CreateSubmissionRequest,
    GradeSubmissionRequest,
    SubmissionListItemResponse,
    SubmissionResponse
} from '../types/submission.types';

export const submissionService = {
    /**
     * Save draft submission
     */
    saveDraft: async (data: CreateSubmissionRequest): Promise<SubmissionResponse> => {
        const response = await api.post<ApiResponse<SubmissionResponse>>('/submissions/draft', data);
        return response.data.data;
    },

    /**
     * Submit exercise
     */
    submit: async (data: CreateSubmissionRequest): Promise<SubmissionResponse> => {
        const response = await api.post<ApiResponse<SubmissionResponse>>('/submissions', data);
        return response.data.data;
    },

    /**
     * Get submission by ID
     */
    getById: async (id: string): Promise<SubmissionResponse> => {
        const response = await api.get<ApiResponse<SubmissionResponse>>(`/submissions/${id}`);
        return response.data.data;
    },

    /**
     * Get submission by exercise and student
     */
    getByExerciseAndStudent: async (exerciseId: string, studentId: string): Promise<SubmissionResponse> => {
        const response = await api.get<ApiResponse<SubmissionResponse>>(
            `/submissions/exercise/${exerciseId}/student/${studentId}`
        );
        return response.data.data;
    },

    /**
     * List submissions by exercise
     */
    listByExercise: async (exerciseId: string): Promise<SubmissionListItemResponse[]> => {
        const response = await api.get<ApiResponse<SubmissionListItemResponse[]>>(
            `/submissions/exercise/${exerciseId}`
        );
        return response.data.data;
    },

    /**
    * Grade submission
    */
    gradeSubmission: async (id: string, data: GradeSubmissionRequest): Promise<SubmissionResponse> => {
        const response = await api.put<ApiResponse<SubmissionResponse>>(
            `/submissions/${id}/grade`,
            data
        );
        return response.data.data;
    }
};

File: backend/src/main/java/com/tutor_management/backend/modules/submission/service/SubmissionService.java
package com.tutor_management.backend.modules.submission.service;

import com.tutor_management.backend.modules.submission.dto.request.CreateSubmissionRequest;
import com.tutor_management.backend.modules.submission.dto.request.GradeSubmissionRequest;
import com.tutor_management.backend.modules.submission.dto.response.SubmissionListItemResponse;
import com.tutor_management.backend.modules.submission.dto.response.SubmissionResponse;

import java.util.List;

/**
 * Service interface for managing student submissions.
 * Handles draft saving, final submission, auto-grading coordination, and tutor evaluation.
 */
public interface SubmissionService {
    
    /**
     * Records a draft state for a student's attempt. 
     * Allows students to save progress without triggering the evaluation phase.
     */
    SubmissionResponse saveDraft(CreateSubmissionRequest request, String studentId);
    
    /**
     * Finalizes a student's attempt. 
     * Triggers the auto-grading sequence for MCQ items and notifies relevant tutors.
     */
    SubmissionResponse submit(CreateSubmissionRequest request, String studentId);
    
    /**
     * Retrieves full submission details including individual answers.
     */
    SubmissionResponse getSubmission(String id);
    
    /**
     * Finds the latest attempt record for a specific student on a given exercise.
     */
    SubmissionResponse getSubmissionByExerciseAndStudent(String exerciseId, String studentId);
    
    /**
     * Aggregates all attempts for a specific exercise (Administrative view).
     */
    List<SubmissionListItemResponse> listSubmissionsByExercise(String exerciseId);
    
    /**
     * Retrieves all historically submitted work for a specific student.
     */
    List<SubmissionResponse> listSubmissionsByStudent(String studentId);
    
    /**
     * Records tutor-provided grades and feedback for essay/written responses.
     */
    SubmissionResponse gradeSubmission(String submissionId, GradeSubmissionRequest request);
}
